### PR TITLE
MR-698 - Normalize postcodes for New and Edited Assets

### PR DESCRIPTION
**Pull request by ange47rm**
Added Postcode Helper file
Added postcode normalization for AddAsset and EditAssetDetailsmethods in DynamoDbGateway

### CODE REVIEW

File: AssetInformationApi/V1/Gateways/DynamoDbGateway.cs [hunk 1]
```diff
@@ -88,6 +94,14 @@ public async Task<Asset> AddAsset(AssetDb asset)
             if (ifMatch != existingAsset.VersionNumber)
                 throw new VersionNumberConflictException(ifMatch, existingAsset.VersionNumber);
 
+            if (assetRequestObject is EditAssetAddressRequest editAddressRequest)
+            {
+                if (PostcodeHelpers.IsValidPostCode(editAddressRequest.AssetAddress.PostCode))
```
Comment (Asura5) [inline AssetInformationApi/V1/Gateways/DynamoDbGateway.cs:99]: You could use && instead of nesting another if

**Review by Asura5 — APPROVED** [review]

File: AssetInformationApi/V1/Gateways/DynamoDbGateway.cs [hunk 2]
```diff
@@ -59,6 +61,11 @@ public async Task<Asset> GetAssetByAssetId(GetAssetByAssetIdRequest query)
         [LogCall]
         public async Task<Asset> AddAsset(AssetDb asset)
         {
+            if (PostcodeHelpers.IsValidPostCode(asset.AssetAddress.PostCode))
+            {
+                asset.AssetAddress.PostCode = PostcodeHelpers.NormalizePostcode(asset.AssetAddress.PostCode);
+            }
```
Comment (LBHCallumM) [inline AssetInformationApi/V1/Gateways/DynamoDbGateway.cs:67]: What if the asset isnt valid? Do we not normalise it?

File: AssetInformationApi/V1/Gateways/DynamoDbGateway.cs [hunk 3]
```diff
@@ -88,6 +94,11 @@ public async Task<Asset> AddAsset(AssetDb asset)
             if (ifMatch != existingAsset.VersionNumber)
                 throw new VersionNumberConflictException(ifMatch, existingAsset.VersionNumber);
 
+            if (assetRequestObject is EditAssetAddressRequest editAddressRequest && PostcodeHelpers.IsValidPostCode(editAddressRequest.AssetAddress.PostCode))
+            {
+                    editAddressRequest.AssetAddress.PostCode = PostcodeHelpers.NormalizePostcode(editAddressRequest.AssetAddress.PostCode);
+            }
```
Comment (LBHCallumM) [inline AssetInformationApi/V1/Gateways/DynamoDbGateway.cs:100]: Same question as above

File: AssetInformationApi/V1/Gateways/DynamoDbGateway.cs (hunk 2 above)
Comment (ange47rm) [reply AssetInformationApi/V1/Gateways/DynamoDbGateway.cs:67]: Not sure how you mean, the added code just looks at whatever is the current postcode that comes within the request and normalizes it.

The way asset validation works is not changed by this PR. What changes would you make?

File: AssetInformationApi/V1/Gateways/DynamoDbGateway.cs (hunk 3 above)
Comment (ange47rm) [reply AssetInformationApi/V1/Gateways/DynamoDbGateway.cs:100]: See reply above

File: AssetInformationApi/V1/Gateways/DynamoDbGateway.cs (hunk 2 above)
Comment (LBHCallumM) [reply AssetInformationApi/V1/Gateways/DynamoDbGateway.cs:67]: In the gateway, you only normalise the postcode if its a valid postcode. But that scenario should never occur if the postcode is validated in the controller. I would remove the check, and just normalize everything

**Pull request merged**